src/core/executor.rs

```diff
@@ -34,33 +73,80 @@ impl Executor {
+    /// So that a single call to `map` does not spawn more than `concurrency_limit` tasks at a time in order to not
+    /// overwhelm the pool by a single operation.
+    pub fn multi_thread_with_concurrency_limit(
+        num_threads: usize,
+        prefix: &'static str,
+        concurrency_limit: usize,
+    ) -> crate::Result<Executor> {
+        let pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(num_threads)
+            .thread_name(move |num| format!("{prefix}{num}"))
+            .build()?;
+        Ok(Executor::ThreadPool {
+            thread_pool: Arc::new(pool),
+            concurrency_limit: Some(concurrency_limit),
+        })
     }
 
     /// Perform a map in the thread pool.
     ///
     /// Regardless of the executor (`SingleThread` or `ThreadPool`), panics in the task
     /// will propagate to the caller.
-    pub fn map<A, R, F>(&self, f: F, args: impl Iterator<Item = A>) -> crate::Result<Vec<R>>
-    where
+    pub fn map<
         A: Send,
         R: Send,
+        AIterator: Iterator<Item = A>,
         F: Sized + Sync + Fn(A) -> crate::Result<R>,
-    {
+    >(
+        &self,
+        f: F,
+        args: AIterator,
+    ) -> crate::Result<Vec<R>> {
         match self {
             Executor::SingleThread => args.map(f).collect::<crate::Result<_>>(),
-            Executor::ThreadPool(pool) => {
+            Executor::ThreadPool {
+                thread_pool,
+                concurrency_limit,
+            } => {
+                let limiter = match concurrency_limit {
+                    Some(limit) => {
+                        if *limit == 0 {
+                            return Err(TantivyError::InternalError(
+                                "Concurrency limit cannot be zero.".to_string(),
+                            ));
+                        }
+                        Some(Arc::new(TaskLimiter::new(*limit)))
+                    }
+                    _ => None,
+                };
+
                 let args: Vec<A> = args.collect();
                 let num_fruits = args.len();
                 let fruit_receiver = {
                     let (fruit_sender, fruit_receiver) = crossbeam_channel::unbounded();
-                    pool.scope(|scope| {
+                    thread_pool.scope(|scope| {
                         for (idx, arg) in args.into_iter().enumerate() {
+                            let limiter = limiter.clone();
+
                             // We name references for f and fruit_sender_ref because we do not
                             // want these two to be moved into the closure.
                             let f_ref = &f;
                             let fruit_sender_ref = &fruit_sender;
                             scope.spawn(move |_| {
+                                if let Some(limiter) = limiter {
+                                    let _permit = limiter.acquire();
+                                }
```

Comment:
This doesn't prevent the task from causing issues with other searchers, the semaphore will park the thread it is being executed on, so if you have a pool with `8` threads, and a concurrency limit `2` and do a search with `8` segments to map, one search will lock out all others until complete, this effectively only limits the maximum number of CPU cores the index reader can max out globally, rather than allow multiple searchers to share the pool without standing on each others toes.